src/usps_processor/parse.clj

```diff
@@ -33,24 +34,117 @@
   "Takes the time in the USPS string format, and a timezone id like
    'America/New_York', and returns a DateTime in UTC for that time."
   [time-string timezone-id]
-  (let [timezone (timezone-id->timezone timezone-id)
-        formatter (format/formatter usps-time-format-pattern timezone)
-        time-in-zone (format/parse formatter time-string)]
-    (time/to-time-zone time-in-zone time/utc)))
+  (try
+    (let [timezone (timezone-id->timezone timezone-id)
+          formatter (format/formatter usps-time-format-pattern timezone)
+          time-in-zone (format/parse formatter time-string)]
+      (time/to-time-zone time-in-zone time/utc))
+    (catch Exception ex
+      (log/warn "Exception parsing scan time" ex))))
+
+(defn datetime->date
+  "Converts a datetime to date, or returns nil if datetime is nil"
+  [datetime]
+  (when datetime
+    (coerce/to-date datetime)))
+
+(defn ->imb-data
+  "Combines the structure-digits and routing code and
+   splits them into the possible mailer-ids and
+   serial numbers"
+  [structure-digits routing-code]
+  (when (and structure-digits
+             (= 20 (count structure-digits)))
+    (imb/split-structure-digits
+     (str structure-digits routing-code))))
+
+(defn valid-number?
+  [count number]
+  (let [pattern (str "\\d{" count "}?")
+        regex (re-pattern pattern)]
+    (not (nil? (re-matches regex number)))))
+
+(defn valid-barcode?
+  [zip]
+  (valid-number? 2 zip))
+
+(defn valid-facility-zip?
+  [zip]
+  (valid-number? 5 zip))
+
+(defn valid-scan-time?
+  [scan-time]
+  (isa? (class scan-time) java.util.Date))
+
+(defn valid-customer-number?
+  [customer-number]
+  (valid-number? 15 customer-number))
+
+(defn valid-number-3?
+  [number]
+  (valid-number? 3 number))
+
+(defn valid-number-6?
+  [number]
+  (valid-number? 6 number))
+
+(defn valid-number-9?
+  [number]
+  (valid-number? 9 number))
+
+(defn valid-routing-number?
+  [number]
+  (and (not (nil? number))
+       (not (nil? (re-matches #"\d*?" number)))
+       (#{0 5 9 11} (count number))))
+
+(def inbound-schema
+  {:barcode (s/pred valid-barcode?)
+   :service (s/pred valid-number-3?)
+   :routing s/Str
+   :customer-number (s/pred valid-customer-number?)})
+
+(def outbound-schema
+  {:barcode (s/pred valid-barcode?)
+   :service (s/pred valid-number-3?)
+   :routing s/Str
+   :9-digit-mailer
+   {:mailer-id (s/pred valid-number-9?)
+    :serial-number (s/pred valid-number-6?)}
+   :6-digit-mailer
+   {:mailer-id (s/pred valid-number-6?)
+    :serial-number (s/pred valid-number-9?)}})
+
+(def scan-schema
+  {:facility-zip (s/pred valid-facility-zip?)
+   :operation-code (s/pred valid-number-3?)
+   :scan-time (s/pred valid-scan-time?)
+   :timezone-id s/Str
+   :imb-data (s/conditional :customer-number
+                            inbound-schema
+                            :9-digit-mailer
```

Comment:
Not really...`:6-digit-mailer` always appears with `:9-digit-mailer`. The whole point of the `:N-digit-mailer` pattern is that USPS provides no way to distinguish if the mailer id was 6 or 9 digits in the first place. So this produces both possibilities and pushes the responsibility for deciding the best fit to ballot-scout.